packages/gatsby/src/utils/node-manifest.ts

```diff
@@ -310,12 +314,41 @@ export async function processNodeManifest(
   const manifestFileDir = path.dirname(manifestFilePath)
 
   await fs.ensureDir(manifestFileDir)
-  await fs.writeJSON(manifestFilePath, finalManifest)
 
-  if (verboseLogs) {
+  const previouslyWrittenNodeManifest = previouslyWrittenNodeManifests.get(
+    inputManifest.manifestId
+  )
+
+  // prefer writing over manifests that have no found page if two manifests have the same manifest ID
+  const shouldWriteManifest =
+    // if we've never written a manifest with this ID to disk
+    !previouslyWrittenNodeManifest ||
+    // or we have but it didn't have a page associated with it.
+    (previouslyWrittenNodeManifest &&
+      previouslyWrittenNodeManifest.foundPageBy === `none` &&
+      // and the current manifest does have a page
+      finalManifest.foundPageBy !== `none`)
+
+  if (shouldWriteManifest) {
+    const mutex = createMutex(
```

Comment:
Sorry @DanielSLew I know I mentioned it before but I meant some kind of hand rolled in memory mutex object (or something like that) for this part. Possibly it's good to use createMutex though 😄

Reply:
Mutex has overhead so you're probably looking for a queue? Try to use fastQ to run things 1 by 1. Or create an inflight map that you can track.

Like
```
const inFlight = new Set();
for loop each manifest
if (inFlight .has('id')) {
  return;
 }
 inFlight.add(id)
 // write to disk
 
 ```

Reply:
@wardpeet it's already running in a queue but since file writing takes some time at scale we're hoping we can keep from limiting the writing concurrency to 1. I have an idea on how we can make this work though with promises since we're already keeping track of what we're writing in a Map, I'll DM you about it @DanielSLew